filter-monitor-config.py

```diff
@@ -0,0 +1,457 @@
+        parser.add_section("STATE")
+        parser["STATE"]['debug'] = 'enabled' if SCRIPT_CONTEXT.debug else 'disabled'
+        mode : str = get_available_write_mode(STATE_FILE)
+        with open(STATE_FILE, mode) as file:
+            parser.write(file)
+        SCRIPT_CONTEXT.debug_message(f"State saved to: {STATE_FILE}")
+    except BaseException as error:
+        SCRIPT_CONTEXT.debug_message(f"Error saving state: {error}")
+
+
+
+
+def load_state():
+    if not os.path.isfile(STATE_FILE): return
+    parser = RawConfigParser()
+    parser.read(STATE_FILE)
+    SCRIPT_CONTEXT.debug = (parser['STATE']['debug'] == 'enabled')
+
+
+
+
+def save_to_file(file_path : str, data : str):
+    mode : str = get_available_write_mode(file_path)
+    with open(file_path, mode) as file:
+        file.write(data)
+
+
+
+
+def load_from_file(file_path : str) -> str:
+    file_path = os.path.abspath(file_path)
+    if not os.path.exists(file_path): raise FileNotFoundError(f"Unable to locate config file: {file_path}")
+    if not os.path.isfile(file_path): raise IsADirectoryError(f"Path is not a file: {file_path}")
+    data : str
+    with open(file_path) as file:
+        data = file.read()
+    return data
+
+
+
+
+def get_available_write_mode(file_path : str):
+    file_path = os.path.abspath(file_path)
+    if os.path.isfile(file_path): return "w"
+
+    parent_dir = os.path.dirname(file_path)
+    if os.path.isdir(parent_dir): return "x"
+        
+    raise FileNotFoundError(f'Requested directory does not exist: {parent_dir}')
+
+
+
+
+def settings_as_dict(settings) -> dict:
+    hostAddress = obs.obs_data_get_string(settings, "_address")
+    port = obs.obs_data_get_int(settings, "_port")
+    obsPassword = obs.obs_data_get_string(settings, "_pass")
+
+    return { "obsHost": f'{hostAddress}:{port}',
+            "obsPassword":obsPassword,
+            "filterlist": get_filters(settings) }
+
+
+
+
+def swing_item_to_dict(swing_array_item) -> dict:
+    obj = obs.obs_data_get_json(swing_array_item)
+    return json.loads(json.loads(obj)["value"])
+
+
+
+
+def swing_array_append_filters(swing_array, filters : list):
+    for filter in filters:
+        try:
+            item_as_json = create_list_item_json(filter["filterName"], filter["sourceName"], filter.get("displayName", None), filter.get("onColor", None))
+            swing_item = obs.obs_data_create_from_json(item_as_json)
+            obs.obs_data_array_push_back(swing_array, swing_item)
+            obs.obs_data_release(swing_item)
+        except:
+            SCRIPT_CONTEXT.debug_message(f"Failed to parse filter: {json.dumps(filter)}")
+
+
+
+
+def int_to_rgb_hex(num : int) -> str:
+    hex_str = hex(num) # 0xAABBGGRR (The obs_data_get_int color format)
+    
+    r = hex_str[8:10]
+    g = hex_str[6:8]
+    b = hex_str[4:6]
+    
+    return "".join(('#', r, g, b))
+
+
+
+
+def create_list_item_json(filterName : str, sourceName : str, displayName : str = None, onColor : str = None) -> str:
```

Comment:
Any reason to build as string here and not just a dict to later json.dumps?

Reply:
This is because there is a JSON string within a JSON string, and the innermost string needs to be formatted with escape characters for OBS to interpret it correctly.